bindings/cpp/session.cpp

```diff
@@ -2287,6 +2341,110 @@ async_iterator_result session::cancel_iterator(const key &id, uint64_t iterator_
 	return iterator(id, data);
 }
 
+async_iterator_result session::server_send(const std::vector<std::string> &keys, uint64_t iflags, const std::vector<int> &groups)
+{
+	if (get_groups().empty()) {
+		async_iterator_result result(*this);
+		async_result_handler<iterator_result_entry> handler(result);
+		handler.complete(create_error(-ENXIO, "server_send: local group list is empty"));
+		return result;
+	}
+
+	if (groups.empty()) {
+		async_iterator_result result(*this);
+		async_result_handler<iterator_result_entry> handler(result);
+		handler.complete(create_error(-ENXIO, "server_send: remote group list is empty"));
+		return result;
+	}
+
+	if (keys.empty()) {
+		async_iterator_result result(*this);
+		async_result_handler<iterator_result_entry> handler(result);
+		handler.complete(create_error(-ENXIO, "server_send: key list is empty"));
+		return result;
+	}
+
+	int local_group = get_groups().front();
+	int err;
+
+	struct la {
+		dnet_addr	addr;
+		int		backend_id;
+		dnet_id		id;
+
+		bool operator<(const la &other) const {
+			int cmp = dnet_addr_cmp(&addr, &other.addr);
+			if (cmp < 0)
+				return true;
+			if (cmp > 0)
+				return false;
+			return backend_id < other.backend_id;
+		}
+	};
+
+	std::map<la, std::vector<dnet_raw_id>> raw_ids;
+	for (auto key = keys.begin(), kend = keys.end(); key != kend; ++key) {
+		la l;
+
+		transform(*key, l.id);
+
+		err = dnet_lookup_addr(get_native(), NULL, 0, &l.id, local_group, &l.addr, &l.backend_id);
+		if (err != 0) {
+			async_iterator_result result(*this);
+			async_result_handler<iterator_result_entry> handler(result);
+			handler.complete(create_error(-ENXIO,
+					"server_send: could not locate backend for requested key %s", key->c_str()));
+			return result;
+		}
+
+		dnet_raw_id raw;
+		memcpy(raw.id, l.id.id, DNET_ID_SIZE);
+
+		auto it = raw_ids.find(l);
+		if (it == raw_ids.end()) {
+			l.id.group_id = local_group;
+			raw_ids[l] = std::vector<dnet_raw_id>({raw});
+		} else {
+			it->second.push_back(raw);
+		}
+	}
+
+	const size_t groups_size = groups.size() * sizeof(int);
+
+	std::list<async_iterator_result> results;
+	for (const auto &it: raw_ids) {
+		auto &id = it.first;
+		auto &ids = it.second;
+
+		const size_t ids_size = ids.size() * sizeof(dnet_raw_id);
+
+		data_pointer data = data_pointer::allocate(sizeof(dnet_server_send_request) + ids_size + groups_size);
```

Comment:
Just a hint: you can use `data_buffer` instead of `data_pointer` here like it is done in [session_indexes.cpp#L111](https://github.com/reverbrain/elliptics/blob/master/bindings/cpp/session_indexes.cpp#L111). It allows to simplify code and to avoid offset calculation with `memcpy`.